test/arborist/build-ideal-tree.js

```diff
@@ -2468,3 +2486,90 @@ t.test('do not ERESOLVE on peerOptionals that are ignored anyway', t => {
     })
   }
 })
+
+t.test('allow ERESOLVE to be forced when not in the source', async t => {
+  const types = [
+    'peerDependencies',
+    'optionalDependencies',
+    'devDependencies',
+    'dependencies',
+  ]
+
+  // in these tests, the deps are both of the same type.  b has a peerOptional
+  // dep on peer, and peer is a direct dependency of the root.
+  t.test('both direct and peer of the same type', t => {
+    t.plan(types.length)
+    const pj = type => ({
+      name: '@isaacs/conflicted-peer-optional-from-dev-dep',
+      version: '1.2.3',
+      [type]: {
+        '@isaacs/conflicted-peer-optional-from-dev-dep-peer': '1',
+        '@isaacs/conflicted-peer-optional-from-dev-dep-b': '',
+      },
+    })
+
+    for (const type of types) {
+      t.test(type, async t => {
+        const path = t.testdir({
+          'package.json': JSON.stringify(pj(type)),
+        })
+        t.matchSnapshot(await printIdeal(path, { force: true }), 'use the force')
+        t.rejects(printIdeal(path), { code: 'ERESOLVE' }, 'no force')
+      })
+    }
+  })
+
+  // in these, the peer is a peer dep of the root, and b is a different type
+  t.test('peer is peer, b is some other type', t => {
+    t.plan(types.length - 1)
+    const pj = type => ({
+      name: '@isaacs/conflicted-peer-optional-from-dev-dep',
+      version: '1.2.3',
+      peerDependencies: {
+        '@isaacs/conflicted-peer-optional-from-dev-dep-b': '',
+      },
+      [type]: {
+        '@isaacs/conflicted-peer-optional-from-dev-dep-peer': '1',
+      },
+    })
+    for (const type of types) {
+      if (type === 'peerDependencies')
+        continue
+      t.test(type, async t => {
+        const path = t.testdir({
+          'package.json': JSON.stringify(pj(type)),
+        })
+        t.matchSnapshot(await printIdeal(path, { force: true }), 'use the force')
+        t.rejects(printIdeal(path), { code: 'ERESOLVE' }, 'no force')
+      })
+    }
+  })
+
+  // in these, b is a peer dep, and peer is some other type
+  t.test('peer is peer, b is some other type', t => {
+    t.plan(types.length - 1)
+    const pj = type => ({
+      name: '@isaacs/conflicted-peer-optional-from-dev-dep',
+      version: '1.2.3',
+      peerDependencies: {
+        '@isaacs/conflicted-peer-optional-from-dev-dep-peer': '1',
+      },
+      [type]: {
+        '@isaacs/conflicted-peer-optional-from-dev-dep-b': '',
+      },
+    })
+    for (const type of types) {
+      if (type === 'peerDependencies')
+        continue
+      t.test(type, async t => {
+        const path = t.testdir({
+          'package.json': JSON.stringify(pj(type)),
+        })
+        t.matchSnapshot(await printIdeal(path, { force: true }), 'use the force')
+        t.rejects(printIdeal(path), { code: 'ERESOLVE' }, 'no force')
+      })
+    }
+  })
```

Comment:
what about a fourth test in which b is an **optional peer dep** of the root asserting that b is not installed at the end?

Reply:
That is covered by https://github.com/npm/arborist/pull/227